src/OrleansCodeGenerator/RoslynCodeGenerator.cs

```diff
@@ -309,12 +312,21 @@ private static GeneratedSyntax GenerateForAssemblies(List<Assembly> assemblies,
                 var assembly = assemblies[i];
                 foreach (var attribute in assembly.GetCustomAttributes<ConsiderForCodeGenerationAttribute>())
                 {
-                    ConsiderType(attribute.Type, runtime, targetAssembly, includedTypes, attribute.GenerateSerializer);
+                    ConsiderType(attribute.Type, runtime, targetAssembly, includedTypes, true);
+                    if (attribute.ThrowOnFailure && !SerializerGenerationManager.IsTypeRecorded(attribute.Type))
+                    {
+                        throw new CodeGenerationException(
+                            $"Found {attribute.GetType().Name} for type {attribute.Type.GetParseableName()}, but code"
+                            + " could not be generated. Ensure that the type is accessible.");
+                    }
                 }
 
+                KnownAssemblyAttribute knownAssemblyAttribute;
+                var treatAsSerializable = knownAssemblyAttributes.TryGetValue(assembly, out knownAssemblyAttribute)
```

Comment:
The naming here is hard. `forceConsiderAsSerializable` is an awkward name, but what if I extract that logic out so that the method becomes this:
```C#
private static void ConsiderType(
    Type type,
    bool runtime,
    Assembly targetAssembly,
    ISet<Type> includedTypes,
    bool considerForSerialization = false)
{
    // The module containing the serializer.
    var module = runtime || type.Assembly != targetAssembly ? null : type.Module;
    var typeInfo = type.GetTypeInfo();

    // If a type was encountered which can be accessed and is marked as [Serializable], process it for serialization.
    if (considerForSerialization)
    {
        RecordType(type, module, targetAssembly, includedTypes);
    }
            
    ConsiderGenericBaseTypeArguments(typeInfo, module, targetAssembly, includedTypes);
    ConsiderGenericInterfacesArguments(typeInfo, module, targetAssembly, includedTypes);

    // Collect the types which require code generation.
    if (GrainInterfaceUtils.IsGrainInterface(type))
    {
        if (Logger.IsVerbose2) Logger.Verbose2("Will generate code for: {0}", type.GetParseableName());
                
        includedTypes.Add(type);
    }
}
```
Which means the call site will look like this:
```C#
var considerForSerialization = knownAssemblyAttributes.TryGetValue(assembly, out knownAssemblyAttribute)
                            && knownAssemblyAttribute.TreatTypesAsSerializable;
foreach (var type in assembly.DefinedTypes)
{
    ConsiderType(type, runtime, targetAssembly, includedTypes, considerForSerialization || type.GetTypeInfo().IsSerializable);
}
```